bundler/helpers/lib/functions/version_resolver.rb

```diff
@@ -0,0 +1,206 @@
+      @gemfile_name = gemfile_name
+      @lockfile_name = lockfile_name
+      @using_bundler_2 = using_bundler_2
+      @dir = dir
+      @credentials = @credentials
+    end
+
+    def version_details
+      setup_bundler
+
+      dep = dependency_from_definition
+
+      # TODO: Rewrite this
+      #
+      # If the dependency wasn't found in the definition, but *is*
+      # included in a gemspec, it's because the Gemfile didn't import
+      # the gemspec. This is unusual, but the correct behaviour if/when
+      # it happens is to behave as if the repo was gemspec-only.
+      if dep.nil? && dependency_requirements.any?
+        return "latest"
+      end
+
+      # Otherwise, if the dependency wasn't found it's because it is a
+      # subdependency that was removed when attempting to update it.
+      return nil if dep.nil?
+
+      # If the dependency is Bundler itself then we can't trust the
+      # version that has been returned (it's the version Dependabot is
+      # running on, rather than the true latest resolvable version).
+      return nil if dep.name == "bundler"
+
+      # If the old Gemfile index was used then it won't have checked
+      # Ruby compatibility. Fix that by doing the check manually (and
+      # saying no update is possible if the Ruby version is a mismatch)
+      return nil if ruby_version_incompatible?(dep)
+
+      details = { version: dep.version }
+      if dep.source.instance_of?(::Bundler::Source::Git)
+        details[:commit_sha] = dep.source.revision
+      end
+      details
+    end
+
+    private
+
+    def using_bundler_2?
+      @using_bundler_2
+    end
+
+    def setup_bundler
+      ::Bundler.instance_variable_set(:@root, dir)
+
+      # TODO: DRY out this setup with Functions::LockfileUpdater
+      return unless credentials
+      credentials.each do |cred|
+        token = cred["token"] ||
+                "#{cred['username']}:#{cred['password']}"
+
+        ::Bundler.settings.set_command_option(
+          cred.fetch("host"),
+          token.gsub("@", "%40F").gsub("?", "%3F")
+        )
+      end
+
+      set_bundler_2_flags if using_bundler_2?
+    end
+
+    def set_bundler_2_flags
+      ::Bundler.settings.set_command_option("forget_cli_options", "true")
+      ::Bundler.settings.set_command_option("github.https", "true")
+    end
+
+    # rubocop:disable Metrics/PerceivedComplexity
+    def dependency_from_definition(unlock_subdependencies: true)
+      dependencies_to_unlock = [dependency_name]
+      dependencies_to_unlock += subdependencies if unlock_subdependencies
+      begin
+        definition = build_definition(dependencies_to_unlock)
+        definition.resolve_remotely!
+      rescue ::Bundler::GemNotFound => e
+        unlock_yanked_gem(dependencies_to_unlock, e) && retry
+      rescue ::Bundler::HTTPError => e
+        # Retry network errors
+        attempt ||= 1
+        attempt += 1
+        raise if attempt > 3 || !e.message.include?("Network error")
```

Comment:
NOTE: These retries are in addition to retries performed by the main process in `SharedBundlerHelpers#in_a_native_bundler_context` - I'm not sure this should be removed as it's pre-existing logic, but it might be worth a comment to acknowledge we are doing some additional low-level retries which might be multiplicative with the main process?

Reply:
Ah yeah looks like it's already handled in `in_a_native_bundler_context` > `RETRYABLE_ERRORS`

Reply:
I think the double retries may have pre-existed in master, I'll go have a look and verify.